File: src/assets/algorithms/algorithms.py
def bubble_sort(array: list):
  swaps = -1

  while swaps != 0:
    swaps = 0
    for i in range(len(array) - 1):
      if array[i] > array[i + 1]:
        tmp = array[i]
        array[i] = array[i + 1]
        array[i + 1] = tmp
        swaps += 1

def selection_sort(array: list):
  for i in range(len(array)):
    minimum = i

    for j in range(i, len(array)):
      if array[j] < array[minimum]:
        minimum = j
    
    tmp = array[minimum]
    array[minimum] = array[i]
    array[i] =  tmp

def insertion_sort(array: list):
  for i in range(1, len(array)):
    # Swap the firt unsorted element until it finds it's place
    for j in reversed(range(i)):
      if array[j + 1] < array[j]:
        tmp = array[j + 1]
        array[j + 1] = array[j]
        array[i] = tmp
      else:
        break


from math import floor

def merge_sort(array: list, start: int = 0, end: int = None):
  if end is None:
    end = len(array)
    
  if start == end - 1: return # sorted
  
  middle = floor((start + end)/2)

  # Sort each part
  merge_sort(array, start, middle)
  merge_sort(array, middle, end)

  tmp, l, r = [], start, middle

  # Merge the two parts
  while l < middle and r < end:
    if array[r] < array[l]:
      tmp.append(array[r])
      r += 1
    else:
      tmp.append(array[l])
      l += 1 

  while l < middle:
    tmp.append(array[l])
    l += 1

  while r < end:
    tmp.append(array[r])
    r += 1

  for i in range(len(tmp)):
    array[start + i] = tmp[i]


array = [4, 3, 1, 6, 2, 0, 5]
merge_sort(array)
print(array)

File: src/assets/algorithms/algorithms.c
#include <stdio.h>
#include <stdlib.h>

void bubbleSort(int n, int array[]);
void selectionSort(int n, int array[]);
void insertionSort(int n, int array[]);
void mergeSort(int n, int array[]);

int main(void) {
  int array[7] = {4, 3, 1, 6, 2, 0, 5}; 
  // printf("%i\n", arr[i]);
  mergeSort(7, array);
  
  for (int i = 0; i < 7; i++) {
    printf("%i ", array[i]);
  }
  printf("\n");
  return 0;
}

void bubbleSort(int n, int array[]) {
  int swaps = -1;

  while (swaps != 0) {
    swaps = 0;
    for (int i = 0; i < n - 1; i++) {
      if (array[i] > array[i + 1]) {
        int tmp = array[i];
        array[i] = array[i + 1];
        array[i + 1] = tmp;
        swaps++;
      }
    }
  }
}

void selectionSort(int n, int array[]) {
  for (int i = 0; i < n; i++) {
    int min = i;

    for (int j = i; j < 6; j++) {
      if (array[j] < array[min]) {
        min = j;
      }
    }

    int tmp = array[min];
    array[min] = array[i];
    array[i] = tmp;
  }
}
// {4, 3, 1, 6, 2, 0, 5} i1 
// {3, 4, 1, 6, 2, 0, 5} i2
void insertionSort(int n, int array[]) {
  
  for (int i = 1; i < n; i++) {
    // Swap the firt unsorted element until it finds it's place
    for (int j = i - 1; j >= 0 || array[j + 1] > array[j]; j--) {
      int tmp = array[j];
      array[j] = array[j + 1];
      array[j + 1] = tmp; 
    }
  }
}

void mergeSort(int n, int array[]) {
  if (n == 1) return; // sorted 
  
  int n1 = n/2;
  int n2 = n - n1;

  // Sort each part
  mergeSort(n1, array);
  mergeSort(n2, array + n1);

  int *tmp = (int *) malloc(n * 4);

  // Merge the two parts
  for (int i = 0, j = 0, k = n1; j <= n1 && k <= n; i++) {
    if ((k == n || array[j] < array[k]) && j != n1) {
      tmp[i] = array[j];
      j++;
    } else {
      tmp[i] = array[k];
      k++;
    }
  }
  
  for (int i = 0; i < n; i++) {
    array[i] = tmp[i];
  }

  free(tmp);
}